lib/waterline/utils/normalize.js

```diff
@@ -50,6 +50,13 @@ var normalize = module.exports = {
       else if (context.attributes[pk].type == 'string') {
         coercePK = function(pk) {return String(pk).toString();};
       }
+      else if (context.attributes[pk].type == 'uuid') {
+        coercePK = function(pk) {
+          var pattern = /[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12}/i
+          var matches = pattern.exec(pk);
+          return matches && matches[0];
```

Comment:
I think this is the right place for this. From what I can tell, the coercion/normalization behavior is supposed to be independent of the higher level model-level validation, because it's on the query criteria, and not the model instance, and gets used on reads (AFAICT there's no validation on reads). `"foo_abc123..."` isn't invalid per-se, it's like passing `"123"` to an integer primary key.

As for errors, I don't think you can have a `null` primary key in PG, but "this should never happen" is a good reason to throw. I believe it'll behave like a WLError being thrown when something goes wrong with the query--otherwise I don't think there are any other errors being thrown in the places where these are being called (e.g. [here](https://github.com/Shyp/waterline/blob/master/lib/waterline/query/finders/basic.js#L28)). I can look a little further into that and see what's normal/expected in the caller.